### PR TITLE
Timers thread managed by EventsExecutor::spin()

### PR DESCRIPTION
**Pull request by mauropasse**

### CODE REVIEW

File: irobot_events_executor/src/rclcpp/executors/events_executor/events_executor.cpp [hunk 1]
```diff
@@ -78,8 +83,18 @@ EventsExecutor::spin()
...
-  timers_manager_->start();
-  RCPPUTILS_SCOPE_EXIT(timers_manager_->stop(); );
+  spin_has_finished_.store(false);
+
+  timers_thread_ = std::thread([this](){ timers_manager_->run_timers(); });
+
+  RCPPUTILS_SCOPE_EXIT(
+      timers_manager_->notify_stop();
+
+      if (timers_thread_.joinable()) {
```
Comment (alsora) [inline irobot_events_executor/src/rclcpp/executors/events_executor/events_executor.cpp:93]: What about moving this to the destructor, thus removing the `spin_has_finished_` bool and the busy loop in the destructor?

File: irobot_events_executor/src/rclcpp/timers_manager.cpp [hunk 2]
```diff
@@ -24,8 +26,14 @@ TimersManager::~TimersManager()
   // Remove all timers
   this->clear();
 
-  // Make sure timers thread is stopped before destroying this object
-  this->stop();
+  // Make sure timers execution is stopped before destroying this object
+  this->notify_stop();
```
Comment (alsora) [inline irobot_events_executor/src/rclcpp/timers_manager.cpp:30]: Now that the thread is managed outside, do we still need this here?

File: irobot_events_executor/include/rclcpp/timers_manager.hpp [hunk 3]
```diff
@@ -150,10 +143,11 @@ class TimersManager
   std::chrono::nanoseconds get_head_timeout();
 
   /**
-   * @brief Function to check if the timers thread is running
-   * @return true if timers thread is running
+   * @brief Implements a loop that keeps executing ready timers.
+   * This function is intended to run in its own thread.
```
Comment (alsora) [inline irobot_events_executor/include/rclcpp/timers_manager.hpp:147]: I don't think that this comment is useful.
Nothing prevents someone to create a timer manager and call this function, without any thread.
We should simply mention that it's a Blocking function

File: irobot_events_executor/src/rclcpp/executors/events_executor/events_executor.cpp (hunk 1 above)
Comment (mauropasse) [reply irobot_events_executor/src/rclcpp/executors/events_executor/events_executor.cpp:93]: Having the thread joining with `timers_thread_.join();` on destructor doesn't guarantee that the `spin()` has finished.
i.e. the call of `RCPPUTILS_SCOPE_EXIT(timers_manager_..)` could end up in seg fault.
Also, if an exception is thrown, the destructor is not called, so `RCPPUTILS_SCOPE_EXIT` should join threads here, right?

File: irobot_events_executor/src/rclcpp/timers_manager.cpp (hunk 2 above)
Comment (mauropasse) [reply irobot_events_executor/src/rclcpp/timers_manager.cpp:30]: Yes I'm not sure this is needed, but as it was the same pattern as the `spin()` thread I added it

File: irobot_events_executor/src/rclcpp/executors/events_executor/events_executor.cpp (hunk 1 above)
Comment (alsora) [reply irobot_events_executor/src/rclcpp/executors/events_executor/events_executor.cpp:93]: I don't know; it feels like we are making things way too complex than they should be and all the synchronization seems brittle.

I would expect the multi-threaded executor to have similar problems, but it looks like it doesn't

**Pull request closed without merging**